### PR TITLE
feat: autoresearch marketplace foundations — worker seller flow, provenance, coordinator fixes

### LINKED ISSUE #264: feat: autoresearch-at-home integration — GPU marketplace + optimized inference selling
## Vision

Three-sided marketplace on obol-stack powered by [autoresearch](https://github.com/karpathy/autoresearch) and its distributed fork [autoresearch-at-home](https://github.com/mutable-state-inc/autoresearch-at-home):

1. **GPU contributors** sell compute time to the autoresearch swarm, paid via x402
2. **Researchers** run distributed experiments across GPU workers, discovering them via ERC-8004
3. **Service builders** take autoresearch-optimized models and sell apps/inference on top via x402

## Context

**Autoresearch** is Andrej Karpathy's autonomous LLM optimization framework — an AI agent iterates on `train.py` (architecture, hyperparams, optimizer), trains for 5 minutes per experiment, measures `val_bpb` (bits per byte), keeps improvements, reverts failures. ~12 experiments/hour, ~100 overnight.

**Autoresearch-at-home** is the SETI@home-style fork: multiple agents on different GPUs collaborating through a shared coordination layer (currently Ensue). Adds claiming, result publishing, global best tracking, collective intelligence.

**Obol-stack** already has the payment and discovery infrastructure: ServiceOffer CRD, x402 ForwardAuth, ERC-8004 registration, buy-side sidecar. The integration connects autoresearch's GPU demand with obol-stack's payment rails.

### Journey 1: GPU Contributor (earn money)
```
Run worker_api.py on bare metal GPU
→ obol sell http gpu-worker --upstream localhost:8080 --per-hour 0.50
→ x402 gates your GPU → researchers pay per-experiment
→ Worker registered on ERC-8004 with OASF skill: machine_learning/model_optimization
```

### Journey 2: Researcher (optimize models)
```
coordinate.py discover → find GPU workers on 8004scan
→ coordinate.py loop train.py → submit experiments through x402
→ Collect best val_bpb model → publish.py → sell optimized inference
→ Provenance (val_bpb, train hash, param count) flows into registration metadata
```

### Journey 3: Service Builder (build apps on optimized models) ⚠️ GAP
```
Take autoresearch-optimized model → build web app (CV enhancer, code reviewer, etc.)
→ obol sell http my-app --upstream localhost:3000 --per-request 0.05
→ Users hit frontend, pay via x402, get the service
```

### Phase 1: Provenance + Skills
- [x] `spec.provenance` field on ServiceOffer CRD (framework, metric, experimentId, trainHash, paramCount)
- [x] `--provenance-file` flag on `obol sell inference` and `obol sell http`
- [x] Provenance injected into `.well-known/agent-registration.json` by monetize.py
- [x] New embedded skill: `autoresearch` (SKILL.md + publish.py + references)
- [x] New embedded skill: `autoresearch-coordinator` (SKILL.md + coordinate.py + references)

### Phase 2: GPU Marketplace
- [x] `worker_api.py` — Flask HTTP API wrapping train.py (`POST /experiment`, `GET /health`, `GET /status`, `GET /best`)
- [x] `Dockerfile.worker` — CUDA 12.4 container for the worker
- [x] Coordinator reimplements Ensue's THINK/CLAIM/RUN/PUBLISH using 8004scan discovery + x402 payments
- [x] GPU workers sold via existing `obol sell http` (no new CRD type needed)

### Design Decision: Bare Metal GPU
k3d doesn't support GPU passthrough. Workers run on the host, obol-stack proxies via `--upstream http://host.k3d.internal:<port>`.

## The Gap: App-on-Top-of-Inference (Journey 3)

**Today** you can sell raw inference (`obol sell inference`) or gate any HTTP service (`obol sell http`). But there's no scaffolding for the common pattern:

> "I want to build a web app that uses an LLM internally and charge users per-use via x402"

For example, a **CV enhancer service**:
- Frontend: upload form for resumes
- Backend: calls the in-cluster LiteLLM with an autoresearch-optimized model
- Payment: x402 gates the whole service per-request

**What's missing:**
1. **App template / scaffold** — no `obol app create` that generates a web app skeleton with LLM backend wired up
2. **Internal LLM routing** — the app needs to call LiteLLM internally (no x402 on internal calls) while the app itself is x402-gated externally
3. **Frontend payment UX** — no x402 payment widget/SDK for browser-based payment flows (today x402 is API-to-API)
4. **Deployment pattern** — no documented pattern for "deploy my app container into the cluster and gate it"

**Proposed solution direction:**
- `obol app create <name> --template inference-app` — scaffolds a Next.js/Flask app with LiteLLM client pre-configured
- App deployed into cluster with internal access to `litellm.llm.svc:4000` (no payment on internal calls)
- `obol sell http <name> --upstream <app-svc>` gates the external-facing endpoint
- x402 browser SDK or payment redirect flow for end-user UX

This needs proper spec work — filing separately or expanding here once we have a concrete prototype (starting with a CV enhancer).

## Phase 3 (Future, not in scope)

- On-chain experiment registry smart contract (experiments, results, bounties)
- GPU metering sidecar (`x402-meter`) for time-based billing
- IPFS/Filecoin model artifact storage (content-addressed by hash)
- Frontend: leaderboard page, experiment dashboard, GPU marketplace view

## Ensue → obol-stack Mapping

| Ensue concept | obol-stack replacement |
|---|---|
| `join_hub()` | `obol sell http` (register GPU as ServiceOffer) |
| `claim_experiment()` | `POST /experiment` to discovered worker via x402 |
| `publish_result()` | ERC-8004 `setMetadata("provenance", {...})` |
| `pull_best_config()` | Query 8004scan API for best val_bpb |
| `ask_swarm()` | Query registered agents' `.well-known` metadata |
| Leaderboard | 8004scan filtered by `machine_learning/model_optimization` skill |

### PR DESCRIPTION
**Pull request by bussyjd**
## Summary

This PR adds the first solid marketplace foundations for autoresearch on obol-stack:

- canonical provenance metadata for optimized models/services
- a sell-side GPU worker flow for paid autoresearch experiments
- richer worker registration metadata for discovery/ranking
- coordinator fixes so discovery/payment work against the current 8004scan and remote-signer APIs
- documentation refresh to match the implemented behavior

The goal is to support three paths over time:

1. GPU contributors selling experiment execution capacity
2. researchers coordinating autoresearch runs across remote workers
3. service builders publishing optimized models and selling inference/apps on top

This PR mainly strengthens the GPU seller and coordinator foundations.

### 1. Canonical provenance support

Adds a consistent provenance shape across:
- `publish.py`
- `--provenance-file`
- `ServiceOffer.spec.provenance`
- stored inference deployment config
- generated `agent-registration.json`

Canonical provenance shape:

```json
{
  "framework": "autoresearch",
  "metricName": "val_bpb",
  "metricValue": "0.9973",
  "experimentId": "abc123def456",
  "trainHash": "sha256:deadbeefcafebabe",
  "paramCount": "50000000"
}
```

Key fixes:
- removed nested `metric` object
- switched to one camelCase schema
- made `trainHash` explicitly `sha256:...`
- made `paramCount` stringified consistently
- made provenance loading strict with unknown-field rejection

### 2. GPU seller flow

Adds a concrete seller-side worker implementation:

- `Dockerfile.worker`
- new embedded skill: `autoresearch-worker`
- `worker_api.py` exposing:
  - `GET /health`
  - `GET /healthz`
  - `GET /status`
  - `GET /best`
  - `GET /experiments/<id>`
  - `POST /experiment`

Worker behavior:
- runs one experiment at a time
- stores results/logs on disk
- extracts `val_bpb`
- computes `train_hash`
- tracks best result
- returns `409` while busy

This gives a concrete path for:
- running a GPU-backed worker
- exposing it behind a Kubernetes Service
- monetizing it via `obol sell http`

### 3. Richer registration metadata for GPU workers

Adds support for passing structured worker metadata at sell time:

```bash
obol sell http autoresearch-worker   ...   --register-metadata gpu=A100-80GB   --register-metadata framework=pytorch   --register-metadata best_val_bpb=1.234   --register-metadata total_experiments=42
```

Changes:
- `cmd/obol/sell.go` adds `--register-metadata key=value`
- CRD adds `spec.registration.metadata`
- `monetize.py` now:
  - includes `registration.metadata` in generated `agent-registration.json`
  - mirrors it into indexed on-chain metadata as `metadata.<key>`

This makes GPU workers easier to discover, compare, and rank.

### 4. Coordinator fixes for current APIs

Fixes the autoresearch coordinator to match the current external APIs:

- 8004scan:
  - uses `/api/v1/public/agents`
  - handles the current `data` response shape
  - prefers `raw_metadata.offchain_content`
  - falls back to the off-chain URI only when needed

- OASF extraction:
  - reads the actual `services[]` entry with `name: "OASF"`

- remote-signer:
  - uses `GET /api/v1/keys`
  - uses `POST /api/v1/sign/<address>/typed-data`

This resolves the previous mismatch where the coordinator used stale 8004scan behavior and old `eth2` signer endpoints.

### 5. Documentation refresh

Refreshes the autoresearch coordinator docs/references so they match the code:
- current 8004scan endpoint shape
- current signer API
- real OASF registration structure
- worker examples using `/services/autoresearch-worker`

## Commits included

- `4b992f5` feat: add autoresearch GPU worker sell-side flow
- `3506ad4` feat: add richer GPU worker registration metadata
- `dc81c85` fix: unify canonical provenance schema
- `2e9a436` fix: align autoresearch coordinator with current APIs
- `71a42d2` docs: refresh autoresearch coordinator references

## Validation performed locally

Validated in this environment:
- Python unit tests for worker helper behavior
- Python unit tests for registration metadata generation
- Python unit tests for provenance generation
- Python unit tests for coordinator helper logic
- `py_compile` on touched Python files
- live `coordinate.py discover --limit 1` against the current 8004scan API

## Notes / scope

This PR improves the marketplace foundations, but it does not yet claim:
- full trustless GPU scheduling
- a complete replacement for Ensue shared-state semantics
- production-complete GPU deployment packaging
- full end-to-end integration coverage for every marketplace journey

It is best understood as:
- a real sell-side worker foundation
- canonical provenance plumbing
- coordinator compatibility fixes
- better discovery metadata

## Follow-up work

Likely next steps after this PR:
- k3s GPU worker chart / deployment packaging
- fuller end-to-end integration tests
- additional worker capability metadata conventions
- deeper coordinator/state model work beyond discovery + payment

Closes #264

### CODE REVIEW

**Comment by bussyjd** [comment]
Superseded by feat/monetize-path integration branch
